### PR TITLE
[luci] Use layer info during quantization

### PR DESCRIPTION
**Pull request by jinevening**
This uses layer info during quantization.

ONE-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

---
Related to: https://github.com/Samsung/ONE/issues/8512
Draft PR: https://github.com/Samsung/ONE/pull/8537

### CODE REVIEW

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 1]
```diff
@@ -896,6 +898,35 @@ void QuantizeWithMinMaxPass::set_output_type(loco::Graph *g) const
...
+    }
+
+    if (not found)
+      throw std::runtime_error("No such layer named " + name +
+                               ". Check layer names in the quantization configuration file.");
+  }
+
+  assert(info_by_name.size() == _ctx->layers_info.size()); // FIX_ME_UNLESS
+
+  return info_by_name;
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:927]: I'm not sure how big size memory would be for return but how about
```cxx
bool QuantizeWithMinMaxPass::create_layer_info_map(loco::Graph *g, LayerInfoMap &info_by_name) const
```
return false if not found or true if OK.

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 2]
```diff
@@ -896,6 +898,35 @@ void QuantizeWithMinMaxPass::set_output_type(loco::Graph *g) const
...
+  for (auto &&info : _ctx->layers_info)
+  {
+    auto name = info.name;
+    bool found = false;
+    for (auto node : loco::active_nodes(loco::output_nodes(g)))
+    {
+      auto cnode = loco::must_cast<luci::CircleNode *>(node);
+      if (cnode->name() == name)
+      {
+        info_by_name[name] = &info;
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:914]: circle graph assumes `name` should be unique. (there is validation but duplicate can splip in)
Q) is there any assumptions in `create_layer_info_map` about the name as this returns `map` ?

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp (hunk 1 above)
Comment (jinevening) [reply compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:927]: AFAIK, std::map (and other std containers) will not be copied due to copy elision.

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp (hunk 2 above)
Comment (jinevening) [reply compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:914]: Hmm.. There is no assumption in `_ctx->layers_info`. I should add a logic to handle duplicate names. Thanks!

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp [hunk 3]
```diff
@@ -896,6 +898,35 @@ void QuantizeWithMinMaxPass::set_output_type(loco::Graph *g) const
...
+        break;
+      }
+    }
+
+    if (not found)
+      throw std::runtime_error("No such layer named " + name +
+                               ". Check layer names in the quantization configuration file.");
+  }
+
+  assert(info_by_name.size() == _ctx->layers_info.size()); // FIX_ME_UNLESS
```
Comment (llFreetimell) [inline compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp:925]: Current logic itself has no problem :)
But if we apply following codes, we may be able to expect :D
- Better readability
- Better performance : O(n^2) --> O(n)

```suggestion
  for (auto &&info : _ctx->layers_info)
    info_by_name[info.name] = &info;

  for (auto node : loco::active_nodes(loco::output_nodes(g)))
  {
    auto cnode = loco::must_cast<luci::CircleNode *>(node);
    if (info_by_name.find(cnode->name()) == info_by_name.end())
    {
      throw std::runtime_error("No such layer named " + cnode->name() +
                               ". Check layer names in the quantization configuration file.");
    }
  }
```

Comment (jinevening) [reply compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp]: Looks much better!

**Review by llFreetimell — APPROVED** [review]
LGTM

**Review by seanshpark — DISMISSED** [review]
LGTM

File: compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp (hunk 3 above)
Comment (jinevening) [reply compiler/luci/pass/src/QuantizeWithMinMaxPass.cpp]: Hmm.. this logic causes an error. It throws an exception whenever a graph has a node not in `info_by_name`, so I roll backed the code.

**Review by seanshpark — APPROVED** [review]
LGTM

**Review by llFreetimell — APPROVED** [review]
LGTM

**Pull request merged**